crates/factors/src/prepare.rs

```diff
@@ -0,0 +1,85 @@
+use std::{any::Any, marker::PhantomData};
+
+use anyhow::Context;
+
+use crate::{AppComponent, Factor, RuntimeFactors};
+
+pub trait FactorInstanceBuilder: Any {
+    type InstanceState;
+
+    fn build(self) -> anyhow::Result<Self::InstanceState>;
+}
+
+pub struct DefaultInstanceBuilder<T>(PhantomData<fn() -> T>);
+
+impl<T: Default + 'static> FactorInstanceBuilder for DefaultInstanceBuilder<T> {
+    type InstanceState = T;
+
+    fn build(self) -> anyhow::Result<Self::InstanceState> {
+        Ok(Default::default())
+    }
+}
+
+pub trait SelfInstanceBuilder: 'static {}
```

Comment:
This and `DefaultInstanceBuilder` are nice, but I'm not convinced they're worth it. It seems like they might make things harder to understand in order to save a few lines of boilerplate...